server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionWorker.java

```diff
@@ -46,25 +46,29 @@ public CompactionWorker(
 
   @Override
   public void run() {
-    while (!Thread.currentThread().isInterrupted()) {
-      AbstractCompactionTask task = null;
-      try {
-        task = compactionTaskQueue.take();
-      } catch (InterruptedException e) {
-        log.warn("CompactionThread-{} terminates because interruption", threadId);
-        return;
-      }
-      if (task != null) {
-        // add metrics
-        CompactionMetricsRecorder.recordTaskInfo(
-            task, CompactionTaskStatus.POLL_FROM_QUEUE, compactionTaskQueue.size());
-        if (task.checkValidAndSetMerging()) {
-          CompactionTaskSummary summary = task.getSummary();
-          CompactionTaskFuture future = new CompactionTaskFuture(summary);
-          CompactionTaskManager.getInstance().recordTask(task, future);
-          task.start();
+    try {
+      while (!Thread.currentThread().isInterrupted()) {
+        AbstractCompactionTask task = null;
+        try {
+          task = compactionTaskQueue.take();
+        } catch (InterruptedException e) {
+          log.warn("CompactionThread-{} terminates because interruption", threadId);
+          return;
+        }
+        if (task != null) {
+          // add metrics
+          CompactionMetricsRecorder.recordTaskInfo(
+              task, CompactionTaskStatus.POLL_FROM_QUEUE, compactionTaskQueue.size());
+          if (task.checkValidAndSetMerging()) {
+            CompactionTaskSummary summary = task.getSummary();
+            CompactionTaskFuture future = new CompactionTaskFuture(summary);
+            CompactionTaskManager.getInstance().recordTask(task, future);
+            task.start();
+          }
         }
       }
+    } catch (Throwable t) {
+      log.error("CompactionWorker.run(), Exception.", t);
     }
```

Comment:
If exception is throw inside the `task.start()`, the compaction thread will still exits and cannot recover. Move the `try-catch` to the inner of the loop.

Reply:
> If exception is throw inside the `task.start()`, the compaction thread will still exits and cannot recover. Move the `try-catch` to the inner of the loop.

We have ever discussed this point internally. 
If we let compact continuing after exception happens, we worry that it will bring risk to damage tsfile files.
It is safe way to print erorr log and interrupt consequent compact.

So, do you still want to continue compacting after exception?

Reply:
Exceptions in the compaction process will be captured and processed internally in the `innerSpaceCompactionTask`and `CrossSpaceCompactionTask`, and there will be no file corruption. Even if an exception is thrown to the `compactionWorker`, it will be an exception elsewhere. So you can place your `try catch` inside.

Reply:
Benson is right, all throwable should be handled correctly by CompactionExceptionHandler.

Reply:
Thank you, 
I chang the try--catch to the inner of the loop.